### PR TITLE
Notify admin about overdoue books via email

### PR DESCRIPTION
**Pull request by lukasz-jaszewski**

### CODE REVIEW

File: cron/src/tests/test_message_service.py [hunk 1]
```diff
@@ -1,49 +1,43 @@
...
-        assert len([
-            m for m in messages
-            if recipient_match(
-                m, 'borrower_name_1', 'borrower_surname_1')]) == 1
+        assert len(messages) == lenght
 
-        assert len([
-            m for m in messages
-            if recipient_match(
-                m, 'borrower_name_2', 'borrower_surname_2')]) == 1
```
Comment (lukasz-jaszewski) [inline cron/src/tests/test_message_service.py:33]: I deleted those parts because I believe they duplicate functionality of `test_compose_messages_to_user`.

File: cron/src/notifications/message_service.py [hunk 2]
```diff
@@ -44,26 +44,25 @@ def compose_user_messages(self, template, books_records):
...
-            html_document = render(template, data)
+        html_document = render(template, data)
 
-            message = EmailMessage()
+        message = EmailMessage()
+        if books_records:
             message['From'] = self.__sender
-            message['To'] = admin
-            message['Subject'] = subject
+            message['To'] = ','.join(getenv('MAIL_ADMINS').split())
```
Comment (rafalskowronski) [inline cron/src/notifications/message_service.py]: Please remove usage of os.getenv() from the method as it makes method harder to test with unit tests.

File: cron/.env [hunk 3]
```diff
@@ -0,0 +1,25 @@
+MAIL_USERNAME=test
```
Comment (rafalskowronski) [inline cron/.env]: Please remove cron/.env file from the repository.
If you need some environment variables setup to make unit tests working, then they are not unit tests anymore. Unit tests should work isolated from your environment.

File: cron/src/tests/test_message_service.py [hunk 4]
```diff
@@ -1,49 +1,43 @@
...
 from notifications.definitions import RecordInfo, BookInfo, BorrowerInfo
 
 
 class TestMessageService():
-    def test_compose_messages_senders(self, test_data):
+    @pytest.mark.parametrize(
+        'test_method, lenght',
+        [('compose_user_messages', 2), ('compose_admin_messages', 1)]
+    )
+    def test_compose_messages_senders(self, test_data, test_method, lenght):
```
Comment (rafalskowronski) [inline cron/src/tests/test_message_service.py]: It should be 'recipients' instead of 'senders' in test name as it tests if recipient in message body is correct.

File: cron/src/tests/test_message_service.py [hunk 5]
```diff
@@ -1,49 +1,43 @@
...
         service = MessageService(sender)
-        messages = list(service.compose_messages(template, test_data))
-
-        def recipient_match(message, recipient_name, recipient_surname):
-            return str(message.get_content()) == '{}, {}\n'.format(
-                recipient_surname,
-                recipient_name)
+        messages = list(getattr(service, test_method)(template, test_data))
 
         for m in messages:
```
Comment (rafalskowronski) [inline cron/src/tests/test_message_service.py]: Please remove logging. It is a leftover from previous implementation.

File: cron/src/tests/test_message_service.py [hunk 6]
```diff
@@ -1,49 +1,43 @@
...
             logging.info(m)
 
-        assert len(messages) == 2
-        assert len([
```
Comment (rafalskowronski) [inline cron/src/tests/test_message_service.py:25]: Why did you remove checking if recipient is correct?

File: cron/src/tests/test_message_service.py [hunk 7]
```diff
@@ -1,49 +1,43 @@
...
-            m for m in messages
-            if recipient_match(
-                m, 'borrower_name_1', 'borrower_surname_1')]) == 1
+        assert len(messages) == lenght
 
-        assert len([
```
Comment (rafalskowronski) [inline cron/src/tests/test_message_service.py:30]: Why did you remove checking if recipient is correct?

File: cron/src/tests/test_message_service.py [hunk 8]
```diff
@@ -1,49 +1,43 @@
...
-    def test_compose_messages_from(self, test_data):
+    @pytest.mark.parametrize(
+        'test_method',
+        ['compose_user_messages', 'compose_admin_messages']
+    )
+    def test_compose_messages_from(self, test_data, test_method):
         sender = 'foo <foo@example.com>'
         service = MessageService(sender)
-        messages = list(service.compose_messages('', test_data))
+        messages = list(getattr(service, test_method)('', test_data))
```
Comment (rafalskowronski) [inline cron/src/tests/test_message_service.py]: Why do parametrize test with function name you want to call and then use reflection to call required function? I would rather have two separate tests with no redundant magic.

File: cron/src/tests/test_message_service.py [hunk 9]
```diff
@@ -55,12 +49,20 @@ def test_compose_messages_to(self, test_data):
             m['To'] == 'borrower_surname_2 ' +
                        'borrower_name_2 <borrower_email_2@example.com>']) == 1
 
+    def test_compose_messages_to_admin(self, test_data):
+        sender = 'foo <foo@example.com>'
+        service = MessageService(sender)
+        messages = service.compose_admin_messages('', test_data)
+
+        for m in messages:
+            assert m['To'] == 'test'
```
Comment (rafalskowronski) [inline cron/src/tests/test_message_service.py]: Where do make an arrangement for 'To' to be 'test'? It it comes from .env file, then it isn't unit test anymore.

File: cron/src/tests/test_message_service.py [hunk 10]
```diff
@@ -1,49 +1,43 @@
...
         sender = 'foo <foo@example.com>'
         template = '{{recipient_surname}}, {{recipient_name}}'
         service = MessageService(sender)
-        messages = list(service.compose_messages(template, test_data))
-
-        def recipient_match(message, recipient_name, recipient_surname):
-            return str(message.get_content()) == '{}, {}\n'.format(
-                recipient_surname,
-                recipient_name)
+        messages = list(getattr(service, test_method)(template, test_data))
```
Comment (rafalskowronski) [inline cron/src/tests/test_message_service.py]: Why do parametrize test with function name you want to call and then use reflection to call required function? I would rather have two separate tests with no redundant magic.

**Review by rafalskowronski — CHANGES_REQUESTED** [review]

File: cron/src/tests/test_message_service.py (hunk 1 above)
Comment (rafalskowronski) [reply cron/src/tests/test_message_service.py:33]: No, they are not.

File: cron/src/run_task.py [hunk 11]
```diff
@@ -53,14 +54,15 @@ def send_notifications():
...
-            try:
-                smtp.send(message)
-            except TypeError:
-                pass
+            records,
+            admin_emails
+        )
+        try:
+            smtp.send(message)
+        except TypeError:
```
Comment (rafalskowronski) [inline cron/src/run_task.py:64]: Why do you catch TypeError here?

**Review by rafalskowronski — APPROVED** [review]
